File: C++/6_13/Retangulo.cpp
#include <iostream>
#include <string>
#include <math.h>

#include "Retangulo.h"

COORDENADA getCoordenada(double x, double y) {
    COORDENADA c;
    c.x = x;
    c.y = y;
    return c;
}

void Retangulo::showMsg(std::string msg) {
    std::cout << BLUE << "[PROCESS] " << NORMAL << spacerString(msg, STRING_SIZE);
}

void Retangulo::showMsg(std::string s, bool v) {
        std::cout << BLUE << spacerString(s, STRING_RESULT_SIZE) << (v ? GREEN : RED) << v << NORMAL << std::endl;
}

void Retangulo::showPassed() {
    std::cout << GREEN << "Passed" << NORMAL << std::endl;
}

void Retangulo::showError() {
    std::cout << RED << "Error" << NORMAL << std::endl;
    error = true;
}

void Retangulo::showError(std::string msg) {
    std::cout << RED << "Error\n" << msg << NORMAL << std::endl;
    error = true;
}

std::string Retangulo::spacerString(std::string s, int size) {
    int s_size = s.size();
    int max = size - s_size;

    if(size > s_size)
        for(int i=0; i < max; i++)
            s += STRING_SPACER;

    return s;
}

// calcula a distância entre dois pontos
double Retangulo::calcDistancia(COORDENADA p1, COORDENADA p2) {
    return sqrt(pow(p1.x - p2.x, 2) + pow(p1.y - p2.y, 2));
}

// organiza os pontos do retângulo, evitando erros nos cálculos
void Retangulo::orderCoordenadas() {
    /* 
    Representação do retângulo para as coordenadas
    
    supEsq (A)  - menor X e maior Y
    maior (B)   - maior X e maior Y
    infDir (C)  - maior X e menor Y
    menor (D)   - menor X e menor Y

    supEsq (A)         maior (B)
    -------------------
    |                 |
    |                 |
    |                 |
    -------------------
    menor (D)        infDir (C)

    */

   int menor, maior, supEsq, infDir;

    double 
        menorX = _coordenadas[0].x, menorY = _coordenadas[0].y,
        maiorX = _coordenadas[0].x, maiorY = _coordenadas[0].y;

    for(int i=0; i < 4; i++) {
        COORDENADA c = _coordenadas[i];
        
        // separando menor
        if(c.x <= menorX && c.y <= menorY) {
            menorX = c.x;
            menorY = c.y;
            menor = i;
        }

        // separando maior
        if(c.x >= maiorX && c.y >= maiorX) {
            maiorX = c.x;
            maiorY = c.y;
            maior = i;
        }
    }

    // separando superior esquerdo e inferior direito
    int ops[2], aux = 0;
    for(int i=0; i < 4; i++) {
        if(i == maior || i == menor)
            continue;
        
        ops[aux++] = i;
    }

    if(_coordenadas[ops[0]].x < _coordenadas[ops[1]].x) {
        supEsq = ops[0];
        infDir = ops[1];
    } else {
        supEsq = ops[1];
        infDir = ops[0];
    }

    // atribuindo valores corretos
    Retangulo::A = _coordenadas[supEsq]; // A
    Retangulo::B = _coordenadas[maior];  // B
    Retangulo::C = _coordenadas[infDir]; // C
    Retangulo::D = _coordenadas[menor];  // D
}

// armazena as coordenadas digitadas pelo usuário
void Retangulo::setCoordenadas() {
    using namespace std;
    COORDENADA cA, cB, cC, cD;

    cout << "\nCoordenadas A \n\tX: ";
    cin >> cA.x;
    cout << "\tY: ";
    cin >> cA.y;

    cout << "\nCoordenadas B \n\tX: ";
    cin >> cB.x;
    cout << "\tY: ";
    cin >> cB.y;

    cout << "\nCoordenadas C \n\tX: ";
    cin >> cC.x;
    cout << "\tY: ";
    cin >> cC.y;

    cout << "\nCoordenadas D \n\tX: ";
    cin >> cD.x;
    cout << "\tY: ";
    cin >> cD.y;

    setCoordenadas(cA, cB, cC, cD);
}

// armazena as coordenadas passadas por parâmetro
void Retangulo::setCoordenadas(COORDENADA A, COORDENADA B, COORDENADA C, COORDENADA D) {
    // verificar se todos os valores pertencem ao primeiro quadrante e são inferiores ou iguais a 20
    _coordenadas[0] = A;
    _coordenadas[1] = B;
    _coordenadas[2] = C;
    _coordenadas[3] = D;

    showMsg("verificando quadrante...");
    if(!checkCoordenadas()) {
        showError("O retangulo nao esta no primeiro quadrante ou possui valores maiores que 20!");
        return;
    }

    primeiroQuadrante = true;
    showPassed();


    showMsg("ordenando coordenadas...");
    orderCoordenadas();
    showPassed();

    showMsg("verificando pontos iguais...");
    if(checkDupliCoordenadas()) {
        showError("O retangulo possui vertices iguais, invalidando sua existencia!");
        return;
    }

    verticesDiferentes = true;
    showPassed();

    isRetangulo();
}
    
// exibe as coordenadas do retângulo
void Retangulo::displayCoordenadas() {
    using namespace std;
    cout << BLUE << "A " << GREEN << "(" << Retangulo::A.x << ", " << Retangulo::A.y << ")" << NORMAL << endl;
    cout << BLUE << "B " << GREEN << "(" << Retangulo::B.x << ", " << Retangulo::B.y << ")" << NORMAL << endl;
    cout << BLUE << "C " << GREEN << "(" << Retangulo::C.x << ", " << Retangulo::C.y << ")" << NORMAL << endl;
    cout << BLUE << "D " << GREEN << "(" << Retangulo::D.x << ", " << Retangulo::D.y << ")" << NORMAL << endl;
}

// verifica se cada coordenada está no primeiro quadrante sem que x ou y seja maior que 20.0
bool Retangulo::checkCoordenadas() {
    for(COORDENADA c: _coordenadas) {
        if(c.x <= 0 || c.x > 20 || c.y <= 0 || c.y > 20)
            return false;
    }
    return true;
}

// verifica se há coordenadas iguais
bool Retangulo::checkDupliCoordenadas() {
    for(int i=0; i < 4; i++) {
        COORDENADA c = coordenadas[i];

        for(int j=0; j < 4; j++) {
            COORDENADA c2 = coordenadas[j];
            if(i == j)
                continue;
            if(c.x == c2.x && c.y == c2.y)
                return true;
        }
    }
    return false;
}

// verifica se forma um retângulo (4 ângulos retos internamente)
bool Retangulo::isRetangulo() {
    double 
        AB = calcDistancia(Retangulo::A, Retangulo::B),
        BC = calcDistancia(Retangulo::B, Retangulo::C),
        CD = calcDistancia(Retangulo::C, Retangulo::D),
        DA = calcDistancia(Retangulo::D, Retangulo::A),
        AC = calcDistancia(Retangulo::A, Retangulo::C), // distancia interna
        BD = calcDistancia(Retangulo::B, Retangulo::D); // distancia interna

    distancias.AB = AB;
    distancias.BC = BC;
    distancias.CD = CD;
    distancias.DA = DA;

    showMsg("verificando angulos internos...");
    double 
        angA = getAngulo(Retangulo::A, Retangulo::B, Retangulo::D),
        angB = getAngulo(Retangulo::B, Retangulo::C, Retangulo::A),
        angC = getAngulo(Retangulo::C, Retangulo::D, Retangulo::B),
        angD = getAngulo(Retangulo::D, Retangulo::A, Retangulo::C);

    // evitando falso positivo por impresição do cálculo (margem de erro)
    if(angA < 90 || angA >= 91 || angB < 90 || angB >= 91 || angC < 90 || angC >= 91 || angD < 90 || angD >= 91) {
        showError("Angulos internos diferentes de 90");
        std::cout << "A: " << angA << std::endl;
        std::cout << "B: " << angB << std::endl;
        std::cout << "C: " << angC << std::endl;
        std::cout << "D: " << angD << std::endl;
        return false;
    }
    showPassed();

    showMsg("verificando se possui lados paralelos...");
    if(AB != CD || DA != BC) { // impossível formar angulo de 90 graus com lados não paralelos
        showError("Nao possui lados paralelos");
        return false;
    }

    ladosParalelos = true;
    showPassed();

    showMsg("verificando se distancias internas sao equivalentes...");
    if(AC != BD) {
        showError("Distancias internas distintas");
        return false;
    }

    distInternasEquivalentes = true;
    showPassed();

    isRet = true;
    return true;
}

// exibe descrição do retângulo
void Retangulo::displayResult() {
    using namespace std;
    cout << boolalpha;
    cout << "\nResultados encontrados" << endl;

    showMsg("Primeiro quadrante:", primeiroQuadrante);
    showMsg("Vertices diferentes:", verticesDiferentes);
    showMsg("Lados paralelos:", ladosParalelos);
    showMsg("Distancias internas equivalentes:", distInternasEquivalentes);
    showMsg("E um retangulo:", isRet);

    // if(!isRet)
    //     return;

    cout << "\nMedidas" << endl;
    cout << BLUE << "Area:      " << GREEN << getArea() << NORMAL << endl;
    cout << BLUE << "Perimetro: " << GREEN << getPerimetro() << NORMAL << endl;

    cout << "\nComprimento dos Lados" << endl;
    cout << BLUE << "AB: " << GREEN << distancias.AB << NORMAL << endl;
    cout << BLUE << "BC: " << GREEN << distancias.BC << NORMAL << endl;
    cout << BLUE << "CD: " << GREEN << distancias.CD << NORMAL << endl;
    cout << BLUE << "DA: " << GREEN << distancias.DA << NORMAL << endl;

    cout << "\nCoordenadas dos vertices" << endl;
    displayCoordenadas();
}

// calcula o angulo dos pontos ABC (angulo de C)
double Retangulo::getAngulo(COORDENADA A, COORDENADA B, COORDENADA C) {
    double
        a = calcDistancia(B, C),
        b = calcDistancia(A, C),
        c = calcDistancia(B, A);

    double angulo = RAD_TO_GRAU(acos((pow(a, 2) - pow(b, 2) - pow(c, 2)) / (-2 * b * c)));
    return angulo;
}

double Retangulo::getPerimetro() {
    return distancias.AB + distancias.BC + distancias.CD + distancias.DA;
}
double Retangulo::getArea() {
    return distancias.CD * distancias.BC;
}

File: Python/6_13/retangulo.py
import math
from typing import Tuple, Iterable

Coordenada = Tuple[float, float]

PI = 3.14159265358979323846
STRING_SIZE = 60
STRING_RESULT_SIZE = 40
STRING_SPACER = ' '

# ANSI colors
BLUE = "\033[34m"
GREEN = "\033[32m"
RED = "\033[31m"
NORMAL = "\033[m"

class Retangulo:
    error = False
    __primeiroQuadrante = False
    __verticesDiferentes = False
    __ladosParalelos = False
    __distInternasEquivalentes = False
    __isRet = False

    __coordenadas:Tuple[Coordenada] = None
    __A:Coordenada = None
    __B:Coordenada = None
    __C:Coordenada = None
    __D:Coordenada = None

    __distancias = dict(AB=0.0, BC=0.0, CD=0.0, DA=0.0)

    def __init__(self, matriz) -> None:
        self.__ordem = 20

    def __getCoorByIndex(self, __index:int) -> Coordenada:
        if __index < 0 or __index > 3:
            raise IndexError("indice fora do range [0:3]")
        values = [self.__A, self.__B, self.__C, self.__D]
        return values[__index]

    def __showMsg(self, msg:str, v:bool=None) -> None:
        if v == None:
            print(BLUE, "[PROCESS]", NORMAL, msg.ljust(STRING_SIZE, STRING_SPACER), end="")
        else:
            color = GREEN if v else RED
            print(BLUE, msg.ljust(STRING_RESULT_SIZE, STRING_SPACER), color, v, NORMAL)

    def __showPassed(self) -> None:
        print(GREEN, "Passed", NORMAL)
    
    def __showError(self, msg:str=None) -> None:
        if msg:
            print(RED, "Error\n", msg, NORMAL)
        else:
            print(RED, "Error", NORMAL)

        self.error = True

    def __calcDistancia(self, p1:Coordenada, p2:Coordenada) -> float:
        """ calcula a distância entre dois pontos """
        return math.sqrt(pow(p1[0] - p2[0], 2) + pow(p1[1] - p2[1], 2))

    def __orderCoordenadas(self) -> None:
        """ organiza os pontos do retângulo, evitando erros nos cálculos """
        """ 
        Representação do retângulo para as coordenadas
    
        supEsq (A)  - menor X e maior Y
        maior (B)   - maior X e maior Y
        infDir (C)  - maior X e menor Y
        menor (D)   - menor X e menor Y

        supEsq (A)         maior (B)
        -------------------
        |                 |
        |                 |
        |                 |
        -------------------
        menor (D)        infDir (C)
        """
        menor, maior, supEsq, infDir = 0, 0, 0, 0

        menorX = self.__coordenadas[0][0]
        maiorX = self.__coordenadas[0][0]
        menorY = self.__coordenadas[0][1]
        maiorY = self.__coordenadas[0][1]

        for i, (x, y) in enumerate(self.__coordenadas):
            # separando menor
            if x <= menorX and y <= menorY:
                menorX = x
                menorY = y
                menor = i
            
            # separando maior
            if x >= maiorX and y >= maiorY:
                maiorX = x
                maiorY = y
                maior = i

        # separando superior esquerdo e inferior direito
        ops = []
        for i in range(4):
            if i not in (maior, menor):
                ops.append(i)
            
        if self.__coordenadas[ops[0]][0] < self.__coordenadas[ops[1]][0]: # se X do primeiro valor que nao foi menor nem maior for menor que X do outro valor que nao foi menor nem maior
            supEsq = ops[0]
            infDir = ops[1]
        else:
            supEsq = ops[1]
            infDir = ops[0]
        
        # atribuindo valores corretos
        self.__A = self.__coordenadas[supEsq]
        self.__B = self.__coordenadas[maior]
        self.__C = self.__coordenadas[infDir]
        self.__D = self.__coordenadas[menor]


    # verifica se cada coordenada está no primeiro quadrante sem que x ou y seja maior que a ordem da matriz
    def __checkCoordenadas(self) -> bool:
        for c in self.__coordenadas:
            if c[0] <= 0 or c[0] > self.__ordem or c[1] <= 0 or c[1] > self.__ordem:
                return False
        return True


    # verifica se há coordenadas iguais
    def __checkDupliCoordenadas(self) -> bool:
        for i in range(4):
            x1, y1 = self.__getCoorByIndex(i)
            for j in range(4):
                x2, y2 = self.__getCoorByIndex(i)
                if i == j:
                    continue
                if x1 == x2 and y1 == y2:
                    return True
        return False


    # verifica se forma um retângulo (4 ângulos retos internamente)
    def __isRetangulo(self) -> bool:
        AB = self.__calcDistancia(self.__A, self.__B)
        BC = self.__calcDistancia(self.__B, self.__C)
        CD = self.__calcDistancia(self.__C, self.__D)
        DA = self.__calcDistancia(self.__D, self.__A)
        AC = self.__calcDistancia(self.__A, self.__C) # distancia interna
        BD = self.__calcDistancia(self.__B, self.__D) # distancia interna

        self.__distancias["AB"] = AB
        self.__distancias["BC"] = BC
        self.__distancias["CD"] = CD
        self.__distancias["DA"] = DA
        self.__distancias["AC"] = AC # distancia interna
        self.__distancias["BD"] = BD # distancia interna

        self.__showMsg("verificando angulos internos...")
        angs = [
            self.getAngulo(self.__A, self.__B, self.__D), # A
            self.getAngulo(self.__B, self.__C, self.__A), # B
            self.getAngulo(self.__C, self.__D, self.__B), # C
            self.getAngulo(self.__D, self.__A, self.__C)  # D
        ]

        for ang in angs:
            # evitando falso positivo por impresição do cálculo (margem de erro)
            if ang < 90 or ang >= 91:
                self.__showError("Angulos internos diferentes de 90 graus")
                print("A:", angs[0])
                print("B:", angs[1])
                print("C:", angs[2])
                print("D:", angs[3])
                return False
        
        self.__showPassed()
        self.__showMsg("verificando se possui lados paralelos...")
        if AB != CD or DA != BC: # impossível formar angulo de 90 graus com lados não paralelos
            self.__showError("Nao possui lados paralelos")
            return False
        
        self.__ladosParalelos = True
        self.__showPassed()
        
        self.__showMsg("verificando se distancias internas sao equivalentes...")
        if AC != BD:
            self.__showError("Distancias internas distintas")
            return False

        self.__distInternasEquivalentes = True
        self.__showPassed()

        self.__isRet = True
        return True


    # armazena as coordenadas
    def setCoordenadas(self, *coordenadas:Coordenada) -> None:
        if not coordenadas:
            coordenadas = []
            for i in ("A", "B", "C", "D"):
                print(f"\nCoordenadas {i}")
                coordenadas.append([float(input("\tX: ")), float(input("\tY: "))])
        
        self.__coordenadas = coordenadas

        self.__showMsg("verificando quadrante...")
        if not self.__checkCoordenadas():
            self.__showError("O retangulo nao esta no primeiro quadrante ou possui valores maiores que 20!")
            return
        
        self.__primeiroQuadrante = True
        self.__showPassed()

        self.__showMsg("ordenando coordenadas...")
        self.__orderCoordenadas()
        self.__showPassed()

        self.__showMsg("verificando pontos iguais...")
        if not self.__checkDupliCoordenadas():
            self.__showError("O retangulo possui vertices iguais, invalidando sua existencia!")
            return
        
        self.__verticesDiferentes = True
        self.__showPassed()

        self.__isRetangulo()


    # exibe as coordenadas do retângulo
    def displayCoordenadas(self) -> None:
        print(BLUE, "A ", GREEN, tuple(self.__A), NORMAL)
        print(BLUE, "B ", GREEN, tuple(self.__B), NORMAL)
        print(BLUE, "C ", GREEN, tuple(self.__C), NORMAL)
        print(BLUE, "D ", GREEN, tuple(self.__D), NORMAL)

    # exibe descrição do retângulo
    def displayResult(self) -> None:
        print("\nResultados encontrados")

        self.__showMsg("Primeiro quadrante:", self.__primeiroQuadrante)
        self.__showMsg("Vertices diferentes:", self.__verticesDiferentes)
        self.__showMsg("Lados paralelos:", self.__ladosParalelos)
        self.__showMsg("Distancias internas equivalentes:", self.__distInternasEquivalentes)
        self.__showMsg("E um retangulo:", self.__isRet)

        # if not isRet:
        #     return

        print("\nMedidas")
        print(BLUE, "Area:      ", GREEN, self.getArea(), NORMAL)
        print(BLUE, "Perimetro: ", GREEN, self.getPerimetro(), NORMAL)

        print("\nComprimento dos Lados")
        print(BLUE, "AB: ", GREEN, self.__distancias["AB"], NORMAL)
        print(BLUE, "BC: ", GREEN, self.__distancias["BC"], NORMAL)
        print(BLUE, "CD: ", GREEN, self.__distancias["CD"], NORMAL)
        print(BLUE, "DA: ", GREEN, self.__distancias["DA"], NORMAL)

        print("\nCoordenadas dos vertices")
        self.displayCoordenadas()

    def getPerimetro(self) -> float:
        return sum(tuple(self.__distancias.values())[:4])

    def getArea(self) -> float:
        return self.__distancias["CD"] *  self.__distancias["BC"]
    
    # calcula o angulo beta dos pontos ABC (angulo de B)
    def getAngulo(self, A:Coordenada, B:Coordenada, C:Coordenada) -> float:
        a = self.__calcDistancia(B, C)
        b = self.__calcDistancia(A, C)
        c = self.__calcDistancia(B, A)

        angRad = math.acos((pow(a, 2) - pow(b, 2) - pow(c, 2)) / (-2 * b * c))

        return (angRad * 180 / PI)